### PR TITLE
SMC data frame compatibility and update Fingerprints

### LINKED ISSUE #166: Check sample order against train order when proposing with NGram
Double check sample order fits the internal train order when proposing new molecules with NGram.

### LINKED ISSUE #177: Default GaussLogLikelihood predict function missing return_std=True
In the tutorial of iQSPR, after executing iQSPR loop, 
prd_mdls["E"].predict(tmp_fps) may return error due to missing "return_std=True"

### PR DESCRIPTION
**Pull request by stewu5**
close #166
close #177 

Add pd.DataFrame compatibility to all classes.
Add three new fingerprints.
Add counting type support to fingerprints.
Fix fp_size to n_bits in RDKitFP.
Update labels of DescriptorFeature to the actual rdkit feature names.

### CODE REVIEW

File: tests/descriptor/test_base_desc.py [hunk 1]
```diff
@@ -173,6 +173,21 @@ def feature_labels(self):
     assert np.alltrue([np.isnan(e[0]) for e in tmp])
 
 
+def test_base_feature_4(data):
+    featurizer = data['featurizer']()
+    tmp_in = pd.DataFrame([[11, 12], [21, 22], [31, 32], [41, 42]])
+    ret = featurizer.transform(tmp_in, target_col=0)
+    assert isinstance(ret, pd.DataFrame)
+    assert (ret.values.ravel() == np.array([11, 21, 31, 41])).all()
```
Comment (TsumiNa) [inline tests/descriptor/test_base_desc.py:181]: Please use `np.all`
https://numpy.org/doc/stable/reference/generated/numpy.all.html

File: xenonpy/descriptor/base.py [hunk 2]
```diff
@@ -91,6 +92,7 @@ def __init__(
         *,
         on_errors: str = 'raise',
         return_type: str = 'any',
+        target_col: str = None,
```
Comment (TsumiNa) [inline xenonpy/descriptor/base.py:95]: Also can be an integer when no column names were specified

File: xenonpy/descriptor/base.py [hunk 3]
```diff
@@ -206,22 +216,36 @@ def fit_transform(self, X, y=None, **fit_params):
             # fit method of arity 2 (supervised transformation)
             return self.fit(X, y, **fit_params).transform(X, **fit_params)
 
-    def transform(self, entries: Sequence, *, return_type=None, **kwargs):
+    def transform(self, entries: Sequence, *, return_type=None, target_col=None, **kwargs):
         """
         Featurize a list of entries.
-        If `featurize` takes multiple inputs, supply inputs as a list of tuples.
+        If `featurize` takes multiple inputs, supply inputs as a list of tuples,
+        or use pd.DataFrame with specified target_col as a list.
```
Comment (TsumiNa) [inline xenonpy/descriptor/base.py:223]:
```suggestion
        or use pd.DataFrame with parameter ``target_col`` to specify the column name(s).
```

File: xenonpy/inverse/base.py [hunk 4]
```diff
@@ -192,35 +211,58 @@ def log_likelihood(self, X, **kwargs):
...
 
         X, _ = self._check_input(X, **kwargs)
-        for k, lls in self.__loglikelihood_sets__.items():
-            if k in kwargs:
-                k = kwargs[k]  # what is this for? even if k not in kwargs... nothing happen right?
-            if k in X:
+        if isinstance(X, list):
+            for k, lls in self.__loglikelihood_sets__.items():
+                # if k in kwargs:
+                #     k = kwargs[k]  # what is this for? even if k not in kwargs... nothing happen right?
```
Comment (TsumiNa) [inline xenonpy/inverse/base.py:217]: this line temporarily replaces the col name in the previous version. In your version, it has no means. 

for example:
```py
self.g1 = ....
self.g1 = ...
```
The group `g1` will try to find the col name **g1** in the input dataframe. When you call `.transform(..., g1=other_key)`, then just in this call, group `g1` will try to find col `other_key`

**Review by TsumiNa — COMMENTED** [review]
1. In the `BaseFeaturizer.transform`, I think we'd better add a checker in case `not 0` to check if the input type is a dataframe when the featurizer use the parallel we provided.

2. You renamed the `tests/datatools/test_splitter.py` to `test_data_select.py`. Because some fixes has been added in this file recently, the rename lead to a conflict. Please name back this file, if you still want to rename it, we can do it after the merging.

3. The `BaseDescriptor` has many modifications, We will face some new errors for a while 😭 

4. You add a new class `BaseLogLikelihoodSet` based on the `BaseDescriptor`, I think the codes are OK, let give more attention to it when we begin to use it

5. Delete any file under `tests/model_dir/`. These files are generated by the unit test and should be deleted automatically when the test is over. I think the reason why they were left here is you ran some test in PyCharm with just a click, the post clean script doesn't work in that case.

File: xenonpy/descriptor/base.py (hunk 2 above)
Comment (stewu5) [reply xenonpy/descriptor/base.py:95]: It does not support int right now because I do not have a check to say if it is int, use "iloc" instead of loc.
I think your BaseDescriptor also required matching the names, so I presume the same here.
Technically, the column names can also be in other types, like int... and allowing specifying columns with column number not column name will make things very messy, I think. I recommend allowing use of column names only.
I put str to try to encourage people using str for column names... I do not know what would be the right way.
For convenience, maybe change to 
target_col: Union[List[str], str, None] = None,
or
is there something like List[column name]?

File: xenonpy/inverse/base.py (hunk 4 above)
Comment (stewu5) [reply xenonpy/inverse/base.py:217]: I still do not quite understand, maybe you can explain to me in person next time.

**Comment by stewu5** [comment]
> 1. In the `BaseFeaturizer.transform`, I think we'd better add a checker in case `not 0` to check if the input type is a dataframe when the featurizer use the parallel we provided.

I agree. If you want me to do it, can you open a new issue after merging this branch to the master?

> 2. You renamed the `tests/datatools/test_splitter.py` to `test_data_select.py`. Because some fixes has been added in this file recently, the rename lead to a conflict. Please name back this file, if you still want to rename it, we can do it after the merging.

I did not rename it. Maybe it was you? I did not touch that part at all (why would I?).

> 3. The `BaseDescriptor` has many modifications, We will face some new errors for a while 😭

Is there any error? I think the change of BaseDescriptor is not so major. Let me know what are you concerns, or you can leave new issues first and I will review it.

> 4. You add a new class `BaseLogLikelihoodSet` based on the `BaseDescriptor`, I think the codes are OK, let give more attention to it when we begin to use it

This is not new. This is from the last update. I only updated this class the same way I updated BaseDescriptor.

> 5. Delete any file under `tests/model_dir/`. These files are generated by the unit test and should be deleted automatically when the test is over. I think the reason why they were left here is you ran some test in PyCharm with just a click, the post clean script doesn't work in that case.

Do I have to do it?

> 6. Seems lack of test [#217 (comment)](https://github.com/yoshida-lab/XenonPy/pull/217#issuecomment-743113287)

Hum... I saw that the 'custom' choice of return_output in BaseFeaturizer is missing some tests. But other than that, the tests are kind of implicitly done. Maybe I will add more tests later in v0.5.1, but do you think it is ok to leave it as is? The overall change is only -0.30%?

File: xenonpy/descriptor/base.py (hunk 2 above)
Comment (TsumiNa) [reply xenonpy/descriptor/base.py:95]: `target_col: Union[List[str], str, None] = None` is the right way.
I agree with you. We should recommend peoples to use str for cols.

**Review by TsumiNa — COMMENTED** [review]
The codebase is ok for merging.

**Comment by TsumiNa** [comment]
> I did not rename it. Maybe it was you? I did not touch that part at all (why would I?).
>
Oh! I remembered that, yes I did. So please rename this file to `test_spliter.py`. I think that will resolve the conflict.

> Is there any error? I think the change of BaseDescriptor is not so major. Let me know what are you concerns, or you can leave new issues first and I will review it.
>
No, for now, this is OK. I said that just based on my experiences, ok for now, error in the future, always 😺 

> This is not new. This is from the last update. I only updated this class the same way I updated BaseDescriptor.
>
Ok, that's great!

> > 1. Delete any file under `tests/model_dir/`. These files are generated by the unit test and should be deleted automatically when the test is over. I think the reason why they were left here is you ran some test in PyCharm with just a click, the post clean script doesn't work in that case.
> 
> Do I have to do it?
>
Yes, you have to. files under `tests/model_dir/`not part of the source code, but has been added to git and committed. This means git will trace these files all the time, but it's not necessary. Please use `git rm -r --cached tests/model_dir/` to untrack them, and delete the files on your machine, then make a new commit for the push. This [link](https://www.codeblocq.com/2016/01/Untrack-files-already-added-to-git-repository-based-on-gitignore/) is a simple detail about `git rm cache`

> > 1. Seems lack of test [#217 (comment)](https://github.com/yoshida-lab/XenonPy/pull/217#issuecomment-743113287)
> 
> Hum... I saw that the 'custom' choice of return_output in BaseFeaturizer is missing some tests. But other than that, the tests are kind of implicitly done. Maybe I will add more tests later in v0.5.1, but do you think it is ok to leave it as is? The overall change is only -0.30%?
>
Yes the overall changes always small because we already have lots of tests. That is not really a critical problem, just for memory.

> > 1. In the `BaseFeaturizer.transform`, I think we'd better add a checker in case `not 0` to check if the input type is a dataframe when the featurizer use the parallel we provided.
> 
> I agree. If you want me to do it, can you open a new issue after merging this branch to the master?
> 
Ok, I will open a new issue for that.

**Review by TsumiNa — APPROVED** [review]
Great works!
